modules/javafx.graphics/src/main/java/com/sun/javafx/css/FixedCapacitySet.java

```diff
@@ -0,0 +1,547 @@
+                if (elements[i].equals(o)) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        @Override
+        public boolean isSuperSetOf(Collection<?> c) {
+            for (int i = 0; i < size; i++) {
+                if (!c.contains(elements[i])) {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        @Override
+        public boolean add(T e) {
+            ensureNotFrozen();
+
+            if (contains(Objects.requireNonNull(e, "e"))) {
+                return false; // already present, set unchanged
+            }
+
+            if (size == elements.length) {
+                throw new IllegalStateException("set is full");
+            }
+
+            elements[size++] = e;
+
+            return true; // not present, set changed
+        }
+
+        @Override
+        public int hashCode() {
+            int h = 0;
+
+            for (int i = 0; i < size; i++) {
+                h += elements[i].hashCode();
+            }
+
+            return h;
+        }
+    }
+
+    /**
+     * A set which can hold a fixed maximum number of elements. This implementation
+     * uses open addressing to handle hash collisions using linear probing. It has a
+     * memory footprint which is much smaller than an equivalent {@code HashSet} but
+     * has worse worst case performance (for modification and contains) when there
+     * are many collisions. Iteration speed will be similar to other array based
+     * collections (which is to say, faster than {@code HashSet}).
+     *
+     * @param <T> the element type
+     */
+    private static final class OpenAddressed<T> extends FixedCapacitySet<T> {
+        private final T[] elements;
+        private final int requestedCapacity;
+        private final int mask;
+
+        private int size;
+
+        @SuppressWarnings("unchecked")
+        private OpenAddressed(int capacity) {
+            this.requestedCapacity = capacity;
+
+            int shift = Integer.SIZE - Integer.numberOfLeadingZeros(capacity * 2 - capacity / 2);
+
+            /*
+             * The shift calculated ensures the elements array's size will be a power
+             * of 2, and ensures that the load factor of this hash map will be roughly
+             * between 0.3 and 0.7; high load factors are detrimental to performance,
+             * while low load factors will consume more memory than necessary.
+             */
+
+            this.elements = (T[]) new Object[1 << shift];
+            this.mask = (1 << shift) - 1;
+
+            /*
+             * Note: the size of the elements array MUST always be greater than the
+             * requested capacity as the contains check relies on there always being
+             * at least one empty bucket.
+             */
+
+            assert elements.length > requestedCapacity : "must have more buckets than capacity";
```

Comment:
This just asserts that the previous three lines of code were correct. Seems a bit excessive to me, if you really need this tested, I'd prefer a unit test.

Reply:
True, but I think it's a lot easier to follow than the logic before it. I see it a bit more as a hint to future maintainers (which might be me) that there are assumptions being made here that code relies on.

Reply:
I understand that, my objection is only with the `assert` keyword. I don't like test code in production code (which this is, it verifies that an assumption holds). I think only unit tests should contain test code, but it's a minor issue and I'll reapprove as-is.

Reply:
I think (and please correct me if I am wrong) that by default java assertions are disabled.  They are only enabled for the tests, see build.gradle :2106 :2628